pkg/controller/keycloak/keycloak_reconciler.go

```diff
@@ -0,0 +1,40 @@
+package keycloak
+
+import (
+	kc "github.com/keycloak/keycloak-operator/pkg/apis/keycloak/v1alpha1"
+	"github.com/keycloak/keycloak-operator/pkg/common"
+	"github.com/keycloak/keycloak-operator/pkg/model/keycloak"
+)
+
+type Reconciler interface {
+	Reconcile(clusterState *common.ClusterState, cr *kc.Keycloak) (common.DesiredClusterState, error)
+}
+
+type KeycloakReconciler struct { // nolint
+}
+
+func NewKeycloakReconciler() *KeycloakReconciler {
+	return &KeycloakReconciler{}
+}
+
+func (i *KeycloakReconciler) Reconcile(clusterState *common.ClusterState, cr *kc.Keycloak) (common.DesiredClusterState, error) {
+	desired := common.DesiredClusterState{}
+	desired = append(desired, i.getKeycloakServiceDesiredState(clusterState, cr))
+	return desired, nil
+}
+
+func (i *KeycloakReconciler) getKeycloakServiceDesiredState(clusterState *common.ClusterState, cr *kc.Keycloak) common.ClusterAction {
+	service := keycloak.Service(cr)
+
+	if clusterState.KeycloakService == nil {
+		return common.GenericCreateAction{
+			Ref: service,
+			Msg: "create keycloak service",
+		}
+	}
+
+	return common.GenericUpdateAction{
+		Ref: service,
```

Comment:
The service would need to have the clusterIP and resourceversion added to it.

Reply:
Could you please explain, why we need both of them (`resourceVersion` and `clusterIP`)? As far as I know, the API server automatically sets the `resourceVersion` and using a Cluster IP as a Service Type is the default one (other types are `NodePort`, `LoadBalancer` and headless - `None`).

Reply:
Unfortunately I don't have the answer, this was brought up by Peter on his PR that for some reason Service resource updates require the ClusterIP and ResourceVersion to be present. 

I have not done any research into this myself. It should be very easy to test though, I will run your operator locally and see if there is any errors.

Reply:
Ok, great! If you encounter any problems - please let me know.

Reply:
Just tested this locally and am getting the following error on the second reconcile loop `{"level":"error","ts":1569320357.343873,"logger":"kubebuilder.controller","msg":"Reconciler error","controller":"keycloak-controller","request":"keycloak/example-keycloak","error":"Service \"keycloak\" is invalid: [metadata.resourceVersion: Invalid value: \"\": must be specified for an update, spec.clusterIP: Invalid value: \"\": field is immutable]","stacktrace":"github.com/keycloak/keycloak-operator/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/Users/dffrench/go/src/github.com/keycloak/keycloak-operator/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/keycloak/keycloak-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/Users/dffrench/go/src/github.com/keycloak/keycloak-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:217\ngithub.com/keycloak/keycloak-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/Users/dffrench/go/src/github.com/keycloak/keycloak-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\ngithub.com/keycloak/keycloak-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/Users/dffrench/go/src/github.com/keycloak/keycloak-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/keycloak/keycloak-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/Users/dffrench/go/src/github.com/keycloak/keycloak-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/keycloak/keycloak-operator/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/Users/dffrench/go/src/github.com/keycloak/keycloak-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"}`

Reply:
Fixed. Thanks!

Reply:
Perfect, tested again and this is working correctly.